### PR TITLE
convert ball placement play to fsm

### LINKED ISSUE #2797: Convert Ball Placement Play into using FSMs
### Description of the task

Ball Placement Play is currently written as Coroutines and we want to move all of our Plays into using FSMs. The new BallPlacementPlay should do and improve the following things:
- Work when placement occurs inside typical Obstacle zones (like the enemy/friendly goal)
- Work when placement occurs normally anywhere in the playing field
- Work when placement occurs at the edges of the field (should be able to retrieve the ball from the outer edges of the field)

You can look at `PenaltyKickPlayFSM`, `ShootOrPassFSM`, `CreaseDefenceFSM` as examples

### Acceptance criteria

- [ ] Convert BallPlacementPlay into a PlayFSM (create `ball_placement_play_fsm.cpp`) and delete the old Ball Placement Test
- [ ] Supplement `ball_placement_play_test.py` with tests to cover the above described gameplay scenarios for ball placement

### PR DESCRIPTION
**Pull request by germchow**
### Description
Conversion of ball placement play to an fsm, including logic changes from robocup 2023. Created a new branch to clean up git history.

### Testing Done
simulated gameplay tests in python and fsm transition test and regular tests in cpp

### Resolved Issues
resolves #2797  

### Length Justification and Key Files to Review

combining James' ball placement test changes into this PR

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ x] **Remove all commented out code**
- [ x] **Remove extra print statements**: for example, those just used for testing
- [ x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp [hunk 1]
```diff
@@ -110,6 +111,39 @@ Point GoalieFSM::findGoodChipTarget(
     return chip_target;
 }
 
+bool GoalieFSM::shouldEvacuateCrease(const Update &event)
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp]: are you also combining the stagnant goalie kick away in this PR as well? Update the PR description if you are.

File: src/software/ai/hl/stp/play/ball_placement/ball_placement_play_test.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        ]
+    ]
+
+    simulated_test_runner.run_test(
+        # setup argument isn't passed to preserve world state from previous test run
+        inv_always_validation_sequence_set=drop_ball_always_validation_sequence_set,
+        inv_eventually_validation_sequence_set=drop_ball_eventually_validation_sequence_set,
+        ag_always_validation_sequence_set=drop_ball_always_validation_sequence_set,
+        ag_eventually_validation_sequence_set=drop_ball_eventually_validation_sequence_set,
+        test_timeout_s=[10],
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement/ball_placement_play_test.py]: can you increase the test timeout here, 10 seconds is a bit short.

File: src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        // if the initial retreat position is out of the field boundary, have it retreat
+        // towards the closest goal
+        if (!contains(world.field().fieldBoundary(), retreat_position))
+        {
+            bool in_friendly_half = contains(world.field().friendlyHalf(), ball_pos);
+            Point closer_goal     = in_friendly_half ? world.field().friendlyGoalCenter()
+                                                 : world.field().enemyGoalCenter();
+            retreat_direction = (closer_goal - ball_pos).normalize();
+            retreat_position =
+                ball_pos + retreat_direction * (0.5 + ROBOT_MAX_RADIUS_METERS);
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp]: convert the distance to a class constant

File: src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    Point ball_pos = event.common.world.ball().position();
+    std::optional<Point> placement_point =
+        event.common.world.gameState().getBallPlacementPoint();
+    std::vector<Robot> robots = event.common.world.friendlyTeam().getAllRobots();
+
+    // see if the ball is at the placement destination
+    if (placement_point.has_value())
+    {
+        return comparePoints(ball_pos, placement_point.value(), 0.05) &&
+               event.common.world.ball().velocity().length() < 0.1;
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp]: so we have a "ball_not_kicked_threshold" defined in ShootOrPassPlayConfig (https://github.com/UBC-Thunderbots/Software/blob/8607d4565479f032a11333eca817026bca9b8873/src/proto/parameters.proto#L229)

Can we move that constant into a new config proto called AiParameterConfig and use that in ShootOrPassPlayFSM and BallPlacementPlayFSM?

File: src/software/ai/navigator/path_planner/global_path_planner_factory.cpp [hunk 5]
```diff
@@ -52,10 +52,9 @@ GlobalPathPlannerFactory::GlobalPathPlannerFactory(
         {
             auto obstacles = obstacle_factory.createStaticObstaclesFromMotionConstraints(
                 static_motion_constraints, field);
-            planners.emplace(std::make_pair(
-                static_motion_constraints,
-                std::make_shared<EnlsvgPathPlanner>(field.fieldBoundary(), obstacles,
-                                                    ROBOT_MAX_RADIUS_METERS)));
+            planners.emplace(std::make_pair(static_motion_constraints,
+                                            std::make_shared<EnlsvgPathPlanner>(
+                                                field.fieldBoundary(), obstacles, 0)));
```
Comment (itsarune) [inline src/software/ai/navigator/path_planner/global_path_planner_factory.cpp]: if you make this change, can you add AVOID_FIELD_BOUNDARY_ZONE to this file https://github.com/UBC-Thunderbots/Software/blob/8607d4565479f032a11333eca817026bca9b8873/src/software/ai/motion_constraint/motion_constraint_set_builder.cpp#L27 and override it for the BallPlacementTactic in the motion_constraint_visitor

**Review by itsarune — COMMENTED** [review]
thanks for cleaning up the robocup stuff. Did you intend to also include the goalie changes in here as well?

File: src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp (hunk 1 above)
Comment (germchow) [reply src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp]: Accidentally included the goalie changes into this PR

I'm gonna create a separate PR for it

**Pull request closed without merging**